Fix from_matrix when argument is already a valid rotation matrix

timethy opened the pull request:
One possibility of fixing issues #627, #1078 by starting from a random rotation.

I'm not too familiar with how this whole Standard: Distribution<Rotation3<T>> works, so feel free to change to something better.

Of course, one downside of this approach is that in the very rare case of running into the bug (if one gets really unlucky), it might disappear when you re-run it.

Hence, another option would be (if one does not always want to start from a new random rotation) to fix the "random" rotation.

An idea could also be to only do this extra check if the output happens to match the first guess (up to ε), and then start from the next "deterministic" random rotation.

Alternatively, one could check if the determinant of the provided matrix is 1, and then use from_matrix_unchecked, however, this would not fix the case, when the input is a scaled "wrong" rotation matrix such as [2, 0, 0, 0, -2, 0, 0, 0, -2]. (This was suggested in https://github.com/dimforge/nalgebra/issues/627)

PS: The suggested method in the paper, of trying different initializations and checking if the frobenius norm changes, only works for non-orthonormal matrices, as all rotation matrices have the same Frobenius norm of √3. One would have to combine it with the latter approach to first detect orientation matrices.

timethy commented:
I changed the method to perturb the rotation after convergence, and if it changes (to smaller norm), it continues from there. This should work for every possible input, and it is deterministic.

Note that the perturbation axes has to change to avoid being stuck in a maximum for ever (with no change in norm). (If e.g. axis = (1, 0, 0) and m = diag(1, -1, -1) it stays stuck starting from rot = diag(1, 1, 1).)

If I understand correctly, there is at least one axes in which the norm changes, so by cycling through three orthogonal axes it should prevent being stuck.

cpg314 commented:
@sebcrozet 
I think this is a super important PR that fixes a critical issue that has already consumed a lot of time from multiple users (#627, @1078).
Any chance this could be merged? I can vouch for the quality of @timethy 's implementation.

timethy commented:
Hi, I created an alternative way of fixing this bug here:
https://github.com/dimforge/nalgebra/pull/1132

Pick your poison -- the other one relies on sqrt() and this one here uses a fixed constant that is somewhat arbitrary.

sebcrozet commented:
@timethy Hey! Thanks. What’s the benefit of using `eps.sqrt()` instead of the arbitrary constant?

timethy commented:
> 

It has the benefit of being very small (pi/8 is not really very small), but being larger than eps.

One could have another small constant, but it should be strictly larger than eps.

Downside is it assumes that 0 < eps < 1, and the latter assumption might not be true, although I don't see a use case of that big of an epsilon.

sebcrozet commented:
@timethy That makes sense thanks. Could you please:
- Apply the `sqrt` change to this PR (on top of the coding-style changes I made here).
- Use the `pi / 8` increment if `eps >= 1`, just to handle this very unlikely case.

timethy commented:
So eps_disturbance needs to be significantly larger then eps to make convergence fast (especially when eps = T::default_epsilon()), so I changed it to eps_disturbance = max(sqrt(eps), eps^2) and added some tests.

sebcrozet commented:
Perfect, thanks!

cpg314 commented:
Thanks, @timethy and @sebcrozet !
